src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,311 @@
+        let port_size   = Vector2::new(port_width, height);
+        let gap_width   = self.gap_width.get();
+
+        // Align shapes along width.
+        let x_start = -width / 2.0 + NODE_SHAPE_PADDING;
+        let x_delta = port_width;
+        for (index, view) in self.ports.borrow().iter().enumerate(){
+            view.display_object().set_parent(&self.display_object);
+
+            let pos_x = x_start + x_delta * index as f32;
+            let pos_y = 0.0;
+            let pos   = Vector2::new(pos_x,pos_y);
+            view.set_position_xy(pos);
+
+            let shape = &view.shape;
+            shape.sprite.size().set(port_size);
+            shape.shape_width.set(width);
+            shape.padding.set(gap_width);
+            shape.offset_x.set(x_delta * index as f32);
+        }
+    }
+
+    fn set_size(&self, size:Vector2<f32>) {
+        self.size.set(size);
+        self.update_shape_layout_based_on_size();
+    }
+}
+
+
+
+// ===================
+// === OutPutPorts ===
+// ===================
+
+/// Implements the segmented output port area. Provides shapes that can be attached to a `Node` to
+/// add an interactive area with output ports.
+///
+/// The `OutputPorts` facilitate the falling behaviour:
+///  * when one of the output ports is hovered, after a set time, all ports are show and the hovered
+///    port is highlighted.
+///  * when a different port is hovered, it is highlighted immediately.
+///  * when none of the ports is hovered all of the `OutputPorts` disappear. Note: there is a very
+///    small delay for disappearing to allow for smooth switching between ports.
+///
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    /// The FRP api of the `OutPutPorts`.
+    pub frp            : Frp,
+        network        : frp::Network,
+        data           : Rc<OutPutPortsData>,
+}
+
+impl OutputPorts {
+    /// Constructor.
+    pub fn new(scene:&Scene, number_of_ports:u32) -> Self {
+
+        let network        = default();
+        let frp            = Frp::new(&network);
+        let data           = OutPutPortsData::new(scene.clone_ref(),number_of_ports);
+        let data           = Rc::new(data);
+
+        OutputPorts {data,network,frp}.init()
+    }
+
+    fn init(mut self) -> Self {
+        self.init_frp();
+        self
+    }
+
+    fn init_frp(&mut self) {
+        let network = &self.network;
+        let frp     = &self.frp;
+        let data    = &self.data;
+
+        frp::extend! { network
+            eval  frp.set_size ((size) data.set_size(size.into()));
+
+            def set_port_size      = source::<(PortId,f32)>();
+            def set_port_sizes_all = source::<f32>();
+
+            def _set_port_sizes = set_port_sizes_all.map(f!([set_port_size,data](size) {
+                let port_num = data.ports.borrow().len();
+                for index in 0..port_num {
+                    set_port_size.emit((index,*size));
```

Comment:
This is antipattern. Do not use map to emit events to frp. This is very error prone. Instead use the `<=` operator to iterate over frp values.

Reply:
```
eval set_port_opacity_all ([set_port_opacity,data](opacity) {
    let port_num = data.ports.borrow().len();
    for index in 0..port_num {
        set_port_opacity.emit((index,*opacity));
    }
});
```
versus
```
port_ix_to_change <= set_port_opacity_all.map(f_!([data] {
    Vec::<usize>::from_iter(0..data.ports.borrow().len())
}));
set_port_opacity_for_port <- all(&port_ix_to_change,&set_port_opacity_all);
eval set_port_opacity_for_port ([set_port_opacity]((index, opacity)) {
   set_port_opacity.emit((*index,*opacity));
});
```
The later seems a lot more complex. Why is the emitting of events in the loop error prone?